watertap/core/membrane_channel_base.py

```diff
@@ -0,0 +1,904 @@
+                == b.properties_interface[t, x].temperature
+            )
+
+    def add_volumetric_flowrate_balance(self):
+        # VOLUMETRIC FLOWRATE
+        @self.Constraint(
+            self.flowsheet().config.time,
+            self.length_domain,
+            doc="Volumetric flow at interface of inlet",
+        )
+        def eq_equal_flow_vol_interface(b, t, x):
+            return (
+                b.properties_interface[t, x].flow_vol_phase["Liq"]
+                == b.properties[t, x].flow_vol_phase["Liq"]
+            )
+
+    def add_concentration_polarization(
+        self,
+        concentration_polarization_type=ConcentrationPolarizationType.calculated,
+        mass_transfer_coefficient=MassTransferCoefficient.calculated,
+    ):
+
+        solute_set = self.config.property_package.solute_set
+        units_meta = self.config.property_package.get_metadata().get_derived_units
+
+        if concentration_polarization_type == ConcentrationPolarizationType.none:
+
+            @self.Constraint(
+                self.flowsheet().config.time,
+                self.length_domain,
+                solute_set,
+                doc="Concentration polarization",
+            )
+            def eq_concentration_polarization(b, t, x, j):
+                return (
+                    b.properties_interface[t, x].conc_mass_phase_comp["Liq", j]
+                    == b.properties[t, x].conc_mass_phase_comp["Liq", j]
+                )
+
+            return
+
+        if concentration_polarization_type not in (
+            ConcentrationPolarizationType.fixed,
+            ConcentrationPolarizationType.calculated,
+        ):
+            raise ConfigurationError(
+                f"Unrecognized concentration_polarization_type {concentration_polarization_type}"
+            )
+
+        self.cp_modulus = Var(
+            self.flowsheet().config.time,
+            self.length_domain,
+            solute_set,
+            initialize=1.1,
+            bounds=(0.9, 3),
+            domain=NonNegativeReals,
+            units=pyunits.dimensionless,
+            doc="Concentration polarization modulus",
+        )
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            self.length_domain,
+            solute_set,
+            doc="Concentration polarization",
+        )
+        def eq_concentration_polarization(b, t, x, j):
+            return (
+                self.properties_interface[t, x].conc_mass_phase_comp["Liq", j]
+                == self.properties[t, x].conc_mass_phase_comp["Liq", j]
+                * self.cp_modulus[t, x, j]
+            )
+
+        if concentration_polarization_type == ConcentrationPolarizationType.calculated:
+            if mass_transfer_coefficient == MassTransferCoefficient.none:
+                raise ConfigurationError()
+
+            # NOTE: calculated concentration polarization is relegated
+            #       to the unit model as it depends on flux
```

Comment:
Ohh interesting, so the concentration polarization equation will live in different places depending on the configuration arguments. I see why that's necessary for initialization, I'm wondering if we always just want the concentration polarization equation on the unit for it to not move based on config arguments.

Reply:
So, we could move the flux variable, but *not* the flux calculation, to the membrane channel. That would allow the concentration polarization equation to move to the membrane channel as well.

Since the flux calculation depends on the state on the "oppose side", (e.g., in the current RO units the permeate stateblock, in future OARO on another membrane channel), it necessarily has to stay on the unit model.

This gets at the heart of the "seam" question. I think adding a flux variable to the channel is a good idea, but of course it's up to the modeller to calculate it in a reasonable fashion.

@adam-a-a what do you think?

Reply:
Well, I am thinking we'll want to have this on the unit model and that it would be better to have `eq_concentration_polarization` in the same place across config types. 

Moreover, `eq_concentration_polarization`, at a minimum, will differ based on the flow direction and whether the channel is facing the active side of the membrane or the porous support layer of the membrane, which I would think is decided at the unit model level. While it'd be nice to have the CP constraint for each channel control volume, I think it'd be ok to let this one slide and have it setup at the unit model level.

Reply:
One thing to keep in mind is the membrane channel could have different build options for these different configurations (e.g., which side of the membrane am I on, which direction is the flow going).

Reply:
OK, so running with that, how about handling a custom relationship for `eq_concentration_polarization`?

Reply:
Who is doing the customizing?

If we put the flux variables in membrane channel, you would add another option there, I think.

Reply:
I was just imagining if someone wanted to use a custom relationship to account for CP, e.g., some empirical relationship related to some particular module geometry. In any case, I suppose we could have an arg for custom_CP_con or something like that for the add_concentration_polarization method.

Reply:
So maybe we should see how things work out if we put the flux vars on the channel, even though they would technically be shared between adjacent channels.

Reply:
> EDIT: Would it make a difference to the developer whether this would need to be applied in the membrane channel or unit model?


Ah, for adding a new one, no.

But, for utilizing an existing one in a new unit model, yes. Instead of copying+pasting, you call the appropriate method (with the appropriate options) on the membrane channel.

This is the crux of what @andrewlee94 is saying above.